unified-scheduler-pool/src/sleepless_testing.rs

```diff
@@ -61,61 +63,97 @@ mod sleepless_testing_real {
                 .into_iter()
                 .chain(check_points)
                 .collect::<Vec<_>>();
-            let check_points_set = check_points.iter().collect::<HashSet<_>>();
-            assert_eq!(check_points.len(), check_points_set.len());
-
             Self {
                 _name: name,
                 check_points,
-                current_check_point: Mutex::new(initial_check_point),
+                current_index: Mutex::new(0),
                 condvar: Condvar::new(),
             }
         }
 
         fn change_current_check_point(&self, anchored_check_point: String) {
-            let Some(anchored_index) = self
-                .check_points
-                .iter()
-                .position(|check_point| check_point == &anchored_check_point)
+            let mut current_index = self.current_index.lock().unwrap();
+
+            let Some(anchored_index) = self.anchored_index(*current_index, &anchored_check_point)
             else {
-                // Ignore unrecognizable checkpoints...
+                trace!("Ignore {} at {:?}", anchored_check_point, current());
                 return;
             };
 
-            let mut current_check_point = self.current_check_point.lock().unwrap();
-
-            let should_change =
-                match anchored_index.cmp(&self.expected_next_index(&current_check_point)) {
-                    Equal => true,
-                    Greater => {
-                        // anchor is one of future check points; block the current thread until
-                        // that happens
-                        current_check_point = self
-                            .condvar
-                            .wait_while(current_check_point, |current_check_point| {
-                                anchored_index != self.expected_next_index(current_check_point)
-                            })
-                            .unwrap();
-                        true
+            let next_index = self.expected_next_index(*current_index);
+            let should_change = match anchored_index.cmp(&next_index) {
+                Equal => true,
+                Greater => {
+                    trace!("Blocked on {} at {:?}", anchored_check_point, current());
+                    // anchor is one of future check points; block the current thread until
+                    // that happens
+                    let wait_result;
+                    (current_index, wait_result) = self
+                        .condvar
+                        .wait_timeout_while(
+                            current_index,
+                            Duration::from_secs(3),
```

Comment:
from interface perspective 3s seems arbitrary and very odd to me.

Reply:
hehe, the 3s is defined as the maximum duration impatient me can wait before hitting Ctrl+C on terminal.

joking aside, it's indeed odd. but the motto of sleepless_testing is faster unit testing. so, still a step forward than the status quo (indefinite blocking), imo.

that said, if changing to 10sec isn't acceptable still. I'll revert this nicety to be added later and land the others be merged in this pr...

Reply:
Yeah imo would rather not clutter this library-esque interface; iirc you said would eventually be splitting this to a separate crate, so hesitant to add any behavior here that we would not want in a separate crate.

Reply:
okay, reverted the timeout part: 124917000bef569ba61566743c7c6e1cb57a4b4c